### ISSUE TITLE
Transaction API error: Transaction already closed: Transaction is no longer valid. Last state: 'Expired'  P2028

### ISSUE THREAD

**Issue by revmischa** [prisma/prisma#13713]
### Bug description

```
"Transaction API error: Transaction already closed: Transaction is no longer valid. Last state: 'Expired'."
```

I'm using 3.14.0
I increased timeouts and these queries generally take ~500ms though they can take longer under load
```ts
{
      maxWait: 20000, // default: 2000
      timeout: 60000, // default: 5000
    }
```

```
{"is_panic":false,"message":"Transaction API error: Transaction already closed: Transaction is no longer valid. Last state: 'Expired'.","meta":{"error":"Transaction already closed: Transaction is no longer valid. Last state: 'Expired'."},"error_code":"P2028","clientVersion":"3.14.0"} 
```

Slack thread: https://prisma.slack.com/archives/CCWDULGUW/p1654457710257189

### How to reproduce

Do a bunch of complex interactive transactions in different lambda functions running in parallel

### Prisma information

3.14.0, in a lambda layer
serverless stack

### Environment & setup

- OS: Lambda
- Database: Postgres
- Node.js version: 14

### Prisma Version

```
prisma                  : 3.14.0
@prisma/client          : 3.14.0
Current platform        : darwin
Query Engine (Node-API) : libquery-engine 2b0c12756921c891fec4f68d9444e18c7d5d4a6a (at node_modules/@prisma/engines/libquery_engine-darwin.dylib.node)
Migration Engine        : migration-engine-cli 2b0c12756921c891fec4f68d9444e18c7d5d4a6a (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine    : introspection-core 2b0c12756921c891fec4f68d9444e18c7d5d4a6a (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary           : prisma-fmt 2b0c12756921c891fec4f68d9444e18c7d5d4a6a (at node_modules/@prisma/engines/prisma-fmt-darwin)
Default Engines Hash    : 2b0c12756921c891fec4f68d9444e18c7d5d4a6a
Studio                  : 0.460.0
```

**Comment by Abdulrahmanelheyb** [comment]
I get the same error. Any resolve ?

**Comment by revmischa** [comment]
This error is causing me a lot of trouble, I'm happy to try to get more information. Is there anything I can do to get some more useful data for debugging it? Would really like to get this resolved one way or another.

**Comment by Abdulrahmanelheyb** [comment]
I tried add max options is work fine and I changing the value based on NODE_ENV

**Comment by revmischa** [comment]
I wonder if this could be related. Could prisma be keeping the transaction open when this error is triggered? I don't know why there are so many processes idle in transaction.

<img width="2354" alt="Screen Shot 2022-06-16 at 10 37 46 PM" src="https://user-images.githubusercontent.com/245131/174232340-da740db0-2860-4096-b0a0-9da5de15bc05.png">

**Comment by manojmpc91** [comment]
Yes this bug keeps troubling and blocks the development. Have tried killing idle processes, disconnecting all open connections, stopping database server but nothing works. Somehow it finds that some transaction was not closed properly.

This issue was closed with https://github.com/prisma/prisma/issues/9584 in version 3.9
I have version 3.14 but the issue still exists.

**Comment by cesarvspr** [comment]
Still exists on 4.0.0

**Comment by felipdc** [comment]
Any update on this?

**Comment by SevInf** [comment]
Excessive CPU usage was fixed in prisma/prisma-engines#3028.
This fix was released in 4.1.0. If you haven't updated yet, could you update to the latest prisma version and report if you are still seeing this?
If you can reproduce it on 4.1.0+ — can you share the schema and code that causes it?

**Comment by manojmpc91** [comment]
Unfortunately two months back we had to shift to pg (node-postgres) from prisma to get rid of the issue.

**Comment by janpio** [comment]
@cesarvspr @felipdc @Abdulrahmanelheyb @revmischa Any help with creating a reproduction of this? We would love to be able to understand this and fix it - but right now we just do not have any code and way to run it that would lead to this error state. Thanks!

**Comment by T3NED** [comment]
Hi, I'm also experiencing this issue. I will attempt to reproduce it in a simplified environment later on today

**Comment by w8ze-devel** [comment]
Hello,

Any update on this ?

We are experiencing this issue very often too. 

What is done is a simple interactive transaction. Sometimes it success, sometimes not without the underlying any change on the underlying data.

Thanks a lot for any help

**Comment by bazzscript** [comment]
Still Experiencing this issue,
happens when I make a request to the DB in immediate succession, that is if there is more than one request at the same time

**Comment by bazzscript** [comment]
> Still Experiencing this issue, happens when I make a request to the DB in immediate succession, that is if there is more than one request at the same time

Solved this,
seems to happen whenever there is a for each loop inside a prisma.$transaction

**Comment by w8ze-devel** [comment]
And how @bazzscript  did you solde this while using a transaction ?

Thanks

**Comment by garrensmith** [comment]
@w8ze-devel could you look at increasing the timeout for your transaction. To me it sounds like your transaction is timing out before all your operations have completed.

**Comment by T3NED** [comment]
@garrensmith When I faced this issue 4 weeks ago, I was using loops of various complexities (which could have been improved overtime). I kept increasing the timeout, but that didn't fix the issue regardless of the timeout.

**Comment by w8ze-devel** [comment]
@garrensmith thanks for your reply. But I can confirm what @T3NED  said about increasing the transaction timeout.
But i continue to randomly experience the problem

**Comment by bazzscript** [comment]
> And how @bazzscript did you solde this while using a transaction ?
> 
> Thanks

An alternative to my logic didn't involve me using loops. I simply took advantage of that alternative. 

For example, I was looping through aItems, converting them to bItems and creating them on the fly. but instead of doing that, i map through aItems , converting it to a list of bItems 

      const bItems = aItems.map((ct) => ({
        bprice: ct.price,
        bproductId: ct.itemId,
        bquantity: ct.quantity,
      }));
      
  and taking that new array of bItems i gave it to prisma to create all of it on the db
  
  tx.basket.update(
            where: {
            id: basket.id,
          },
          data: {
            bItems: {
              create: bItems,
            },
  )
  
  
  this one is more efficient than what i was doing before where i was looping through items, converting it to bitems and creating it one at a time

**Comment by bazzscript** [comment]
I  generally try to write loops myself as rarely as possible, and this is one of those opportuinity

**Comment by arimendelow** [comment]
+1 on that - sometimes you do need to create objects one at a time, maybe because they refer to each other, and so you can't get away with not using loops

**Comment by hironow** [comment]
I also encountered this issue in ver 4.5.0 and it is happening under high load in implementations where the only way is to use a for loop.

**Comment by heymartinadams** [comment]
I’m wondering if this https://github.com/prisma/prisma/issues/16050 is caused by the same underlying issue. @hironow, interestingly enough, the bug I mention (`Transaction not found`) is also produced inside a `for` loop: https://github.com/ConsciousApps/replicache-demo/blob/main/utils/api/mutations/index.js#L9-L40

**Comment by janpio** [comment]
Can anyone help us with a minimal reproduction? As soon as we have that, we can start into looking what is causing this - but for now we have not found any way to trigger this ourselves so would just be stabbing at it with no way to see if we did anything useful.

**Comment by wezooyuri** [comment]
I don't have a limited reproduction, but I do have some input;

We have seen this exact issue happen slightly after the default timeout of 5s for a transaction. I'm unsure why our (usually) fast queries take as much as 5 seconds to execute and therefore timeout. That seems to be part of this issue. After I restarted the container using the connection the problem resolved (and has been running stable for a few days now).

**Comment by janpio** [comment]
Hey @hironow @heymartinadams @arimendelow @felipdc @cesarvspr @bazzscript @Abdulrahmanelheyb @T3NED @w8ze-devel @wezooyuri @manojmpc91 - this is one of our last blocker for getting `interactiveTransactions` into General Availability. This issue, not having a reproduction or any information about what really is going on, is of course quite a risk for us. Can you please help us understand what is happening here?

- Are you still getting this error message?
- Is this reproducible for you?
- Can you help us understand when this happens and what to do about it?

**Comment by bazzscript** [comment]
Hey @janpio, 
This issue is very tricky to reproduce.
But the common things among all of us that have experience this error, is using a long running for loops

**Comment by conorwaltersUno** [comment]
A simple example would be:

```
const upsertIssue = await prisma.$transaction(async (prismaTransaction) => {
   const upsertApplication = await upsertApplicationData(prismaTransaction, application, issue);
   //the transaction would be closed due to the for loop in the upsertApplicationWithIssueData function so any subsequent upsert function would fail with the error
   const upsertIssue = await upsertIssue(prismaTransaction, application, issue);
 }
 
 async function upsertApplicationData(prismaTransaction, application, Issue) {
  for (const releaseVersion of Issue.ReleaseVersions) {
    const applicationVersion = await prismaTransaction.application.upsert({
      where: {
          application_id: application.id,
        },
      },
      create: {
        application_id: upsertApplication.application_id,
        jira_release_version: issue.releaseVersion,
      },
      update: {
        jira_release_version: issue.releaseVersion,
      },
    });
  }
}
``` 
The upsertApplicationData function called within the transaction is causing this transaction to close with the error:
 
 `Transaction API error: Transaction already closed: A query cannot be executed on a closed transaction..`

**Comment by aqrln** [comment]
@bazzscript 

> But the common things among all of us that have experience this error, is using a long running for loops.

Long running for loops sounds like something that could legitimately take more than the default timeout — hard to tell though without knowing how long exactly is "long".

I'm also curious, are you using a remote database?

@conorwaltersUno in your example, how long does a single `upsert` take and how many `Issue.ReleaseVersions` do you have?

**Comment by bazzscript** [comment]
@aqrin long-running loops can have different "how long?". Lenght of the Item I was looping through is not constant, it could be 10,100, 20 e.t.c and the bigger the items, the big chance of the error coming up.

Am using a local db

**Comment by Pompedup** [comment]
I had the same issue and I solved it with a Promise.all
I work on a local postgre.
I was doing

```ts
const documentsToArchive: DocumentAttributes[] = [...]

await prisma.$transaction(async (trx) => {
  for (const documentToArchive of documentsToArchive) {
    await trx.documents.updateMany({ data: { status: 'ARCHIVED'}, where: documentToArchive })
  }
})
```

Now I do

```ts
const documentsToDelete: DocumentAttributes[] = [...]

await prisma.$transaction(async (trx) => {
  await Promise.all(documentsToArchive.map((documentToArchive) =>
    trx.documents.updateMany({ data: { status: 'ARCHIVED'}, where: documentToArchive })))
})
```

Hope it can help to reproduce & to solve for some people ✌🏻

**Comment by heymartinadams** [comment]
@Pompedup you can keep the `for` loop if you use `for await (const ...`.

**Comment by hironow** [comment]
Thank you so much for proceeding!

my commented

> I also encountered this issue in ver 4.5.0 and it is happening under high load in implementations where the only way is to use a for loop.

I have checked that the loop rewritten to `while` also reproduced the problem.

**Comment by aqrln** [comment]
@Pompedup your example also hints at the transaction actually timing out, since the problem was solved by rewriting a slow sequential algorithm to a faster concurrent one.

Not taking the specific of your query into the account and just talking about the fact that you don't have any dependencies between your queries, I think a slightly better option would've been using a batch transaction instead of an interactive transaction:

```ts
await prisma.$transaction(documentToArchive.map((documentToArchive) =>
  prisma.documents.updateMany({ data: { status: 'ARCHIVED'}, where: documentToArchive }))
```

However, regardless of the type of transaction, the big problem here is `updateMany` inside the loop, since you are probably scanning the whole table on every iteration.

The best option would be getting rid of both the transaction and the loop altogether and replacing it with a single `updateMany` call if possible — this could be especially cheap if you know the IDs of the documents.

@heymartinadams:
> you can keep the for `loop` if you use `for await (const ....`

I don't think this is correct. An array is not an `AsyncIterator` in JavaScript, and using `for await` wouldn't make anything faster anyway since it is sequential too.

**Comment by aqrln** [comment]
Hey everyone!

This is a complex issue in a sense that people run into it under different circumstances, and whether it is easily reproducible or only occurs intermittently or even once also differs for various people.

From what I can see, in most cases where some kind of reproduction was provided, the error looks very plausible and doesn't seem to be an issue in Prisma looking at the code alone without specific numbers.

I believe that in those cases where we couldn't see a reproduction, even where the error was very elusive and hard to reproduce again for the reporter, the error was still technically correct and the transactions timed out. However, there is still an open question of whether it is expected for the _operations inside the transaction_ to take as much time as they took for everyone who reported it — as well as what exactly the numbers were. I think some people might just be facing network issues that make their queries take longer than usual to get to the database (the original report, for example, mentions serverless functions, which implies remote database), but there could theoretically be a possibility that we might have a performance issue somewhere which makes some operation slower when used inside a transaction, although I haven't seen an indication of it so far.

It also looks like some people were just confused by the error itself and thought it was some internal error and not a user-facing error indicating that the transaction has timed out.

To alleviate these problems, we would like to improve the error messages related to interactive transactions to make them more informative and user-friendly, re-add the information that was lost since version 4.1.0, as well as provide new information and more visibility into the expired transactions.

Aside from the reason why the transaction was closed, expired transactions will return the transaction timeout and the actual time that has passed since the start of the transaction as a part of the error message: `Transaction API error: Transaction already closed: A query/commit/rollback cannot be executed on an expired transaction. The timeout for this transaction was X ms, however Y ms passed since the start of the transaction. Consider increasing the interactive transaction timeout.`

Hopefully this should be less confusing and more helpful for those who are doing a lot of work in the transaction callback and run into this error as expected.

For those who don't believe they should be running into this error, it should provide more context about your circumstances that you could share with us — and I would suggest opening new, more specific, issues instead of reusing this very general and ambiguous one. Any reports about this functionality not working as you would expect are still very welcome and wanted!

While the new error message will provide some basic visibility into the timing of an interactive transaction (which is already immensely more useful that what we have now), I would suggest using [tracing](https://www.prisma.io/docs/concepts/components/prisma-client/opentelemetry-tracing) for those who want more advanced visibility and to learn why exactly a transaction is taking as much time as it does. We would really appreciate if you include this information in your bug reports btw!

**Comment by andrewgreenh** [comment]
Just a thought regarding the proposed error message: Would it be good to include a link or a hint how and where to increase the transaction timeout?

**Comment by bazzscript** [comment]
> Just a thought regarding the proposed error message: Would it be good to include a link or a hint how and where to increase the transaction timeout?

Yes, definitely.

**Comment by aqrln** [comment]
@andrewgreenh 

> Just a thought regarding the proposed error message: Would it be good to include a link or a hint how and where to increase the transaction timeout?

We could add a link to the docs there, this is a nice idea. Sorry I didn't see your comment earlier.
Is the current error message good enough for now, or do you think it's not that helpful without a link to the docs?

Optimally we'd also need to add a subheader we could link to directly, as currently the nearest header we could link to is https://www.prisma.io/docs/concepts/components/prisma-client/transactions#interactive-transactions-in-preview, and one would need to scroll a bit until the transaction options.

**Comment by BCsabaEngine** [comment]
My experience with 4.7.1 client: in transaction context (in callback of $transaction fn) if I create a for cycle, the ```await tx.table.create()``` closes the transaction (it does not detect that it in transaction) but without await it works.

**Comment by janpio** [comment]
Can you please open a new issue for this @BCsabaEngine, we want to take a look at this behavior. Thanks.

**Comment by BCsabaEngine** [comment]
> Can you please open a new issue for this @BCsabaEngine, we want to take a look at this behavior. Thanks.

Issue is not issue. Transation timeout occurs problem. Solved.

**Comment by steven-tey** [comment]
Ran into this issue today as well – turns out the fix was to convert my `arr.forEach(...)` loop into a good old `for (let i = 0; i < arr.length; i++)` and everything worked like a charm!

**Comment by Karl-EdwardFPJeanMehu** [comment]
I recently started to experience this. I confirm @steven-tey 's solutions works

**Comment by lamnguyencse17** [comment]
It seems that Prisma does not like loops inside its code I think

```
  return tx.invitationDishOption.createMany({
    data: restaurant.dish.flatMap((dish) =>
      dish.dishOption.map((dishOption) => ({
        id: dishOption.id,
        dishId: dishOption.dishId,
        optionId: dishOption.optionId,
        invitationRestaurantId,
      }))
    )
```

By moving the `flatMap` outside it resolves the problem

**Comment by erdinc** [comment]
after reading provided solutions and comments, I didn't understand why it works with good old for loop but not with `Array.forEach` or `Array.map` and also some people mentioned that using `Promise.all` solved the issue for their use cases. Can someone clarify the reason and work around for that case? it will be nice for people coming afterwards. 

one feedback, I would expect to see more comments from Prisma team. It will give a bit more relief to people using Prisma on their production environment as database layer libraries are one of the core and fundamental parts of most web/server based applications.

**Comment by janpio** [comment]
If you can still reproduce some part of this problem, please open a new issue and provide all the information the bug issue template asks for - and optimally a reproduction we can run ourselves. Thanks. 

For us the issue was closed by improving the error message, as there was no clear reproduction of what was going on. We unfortunately can not possibly respond to all not directly related comments in all our issues. That would make things a lot more confusing to everyone (and impossible for us to manage). Hope you understand.

**Comment by patriciorasquetti** [comment]
I know it's late, but in case you're using prisma transaction inside a arr.foreach() it works if you not use the transaction refence I mean this:
instead of this (see the tx inside the loop):
const result = await prisma.$transaction(async (tx) => {
        deal.activities.forEach(async (activity) => {
                        const dealActivity = await tx.crmDealActivity.create()
        }
}

use this ( ses the prisma inside the loop, maybe it does not recognise the scope)
const result = await prisma.$transaction(async (tx) => {
        deal.activities.forEach(async (activity) => {
                        const dealActivity = await prisma.crmDealActivity.create()
        }
}

it worked for me.
Regards.

**Comment by sebmellen** [comment]
Is there any way to increase the timeout? This is a hard blocker for us right now.

**Comment by ByBogon** [comment]
@sebmellen 

`await prisma.$transaction(
  async (tx) => {
    // Code running in a transaction...
  },
  {
    maxWait: 5000, // default: 2000
    timeout: 10000, // default: 5000
    isolationLevel: Prisma.TransactionIsolationLevel.Serializable, // optional, default defined by database configuration
  }
)`

This is from the official document.
https://www.prisma.io/docs/concepts/components/prisma-client/transactions

**Comment by chrisui** [comment]
I think most people are encountering this issue through running concurrent interactive transactions in a `.map()` and `.forEach()` - it is desireable to be able to do this so would welcome an extension to the documentation to better describe how to achieve concurrent running interactive transactions.

**Comment by notflip** [comment]
> I had the same issue and I solved it with a Promise.all I work on a local postgre. I was doing
> 
> ```ts
> const documentsToArchive: DocumentAttributes[] = [...]
> 
> await prisma.$transaction(async (trx) => {
>   for (const documentToArchive of documentsToArchive) {
>     await trx.documents.updateMany({ data: { status: 'ARCHIVED'}, where: documentToArchive })
>   }
> })
> ```
> 
> Now I do
> 
> ```ts
> const documentsToDelete: DocumentAttributes[] = [...]
> 
> await prisma.$transaction(async (trx) => {
>   await Promise.all(documentsToArchive.map((documentToArchive) =>
>     trx.documents.updateMany({ data: { status: 'ARCHIVED'}, where: documentToArchive })))
> })
> ```
> 
> Hope it can help to reproduce & to solve for some people ✌🏻

How can this be used if the output from the previous record is needed for the next one? For example:

```
await prisma.$transaction(async (tx) => {
        await Promise.all(parsedData.appls.map((appl: any) => {

            const applInsert = tx.appl.create({
                data: ...
            })

            for (const hl01 of appl.hl01s) {
                const hl01Insert = await tx.lONSTRKWN_Hl01.create({
                    data: {
                        ...hl01WithoutChildren,
                        appl_id: applInsert.id
                    }
                })
            }
        }))
}
```

**Comment by notflip** [comment]
> Ran into this issue today as well – turns out the fix was to convert my `arr.forEach(...)` loop into a good old `for (let i = 0; i < arr.length; i++)` and everything worked like a charm!

**Why does this work?** Man I've been breaking my head about this issue, it's also only an issue on a Windows machine, on my mac it's working fine using for...of

**Comment by shadoath** [comment]
I also found that this error would come up when I had a transaction that called an awaited promise to Prisma that was NOT using the transaction.

**Comment by m1212e** [comment]
In my case this error appeared because I fired but did not await a promise from inside the async transaction callback

**Comment by Eh-haque** [comment]
> I had the same issue and I solved it with a Promise.all
> I work on a local postgre.
> I was doing
> 
> ```ts
> const documentsToArchive: DocumentAttributes[] = [...]
> 
> await prisma.$transaction(async (trx) => {
>   for (const documentToArchive of documentsToArchive) {
>     await trx.documents.updateMany({ data: { status: 'ARCHIVED'}, where: documentToArchive })
>   }
> })
> ```
> 
> Now I do
> 
> ```ts
> const documentsToDelete: DocumentAttributes[] = [...]
> 
> await prisma.$transaction(async (trx) => {
>   await Promise.all(documentsToArchive.map((documentToArchive) =>
>     trx.documents.updateMany({ data: { status: 'ARCHIVED'}, where: documentToArchive })))
> })
> ```
> 
> Hope it can help to reproduce & to solve for some people ✌🏻
>

**Comment by c-andrey** [comment]
I solved using tthe promise.all as @Eh-haque commented above

**Comment by avarayr** [comment]
For anyone trying add a LOT of data (for example, during seeding, more than 10,000 rows, Prisma will still timeout even with Promise.all) 
In that case, you can split up the data into chunks of transactions

✅ Works under 500ms to add over 10k rows..

```typescript

    const chunkSize = 100; // I found 100 to work the fastest
    const chunks = [];
    for (let i = 0; i < seed_data.length; i += chunkSize) {
      chunks.push(seed_data.slice(i, i + chunkSize));
    }

    await Promise.all(
      chunks.map(async (chunk) => {
        await db.$transaction(async (tx) => {
          await Promise.all(
            chunk.map((chunkData) => {
              i++;
              return tx.YOUR_TABLE_HERE.create({
                data: chunkData,
              });
            }),
          );
        });
      }),
    );
```

**Comment by raymondKelly** [comment]
I actually suspect the issue could be that people are running async calls in parallel with both of them updating the database at some point in their call cycle. A suspicion I have for why I see this issue is that I am using Promise.all([doAction1, doAction2]) and at some point action1 is timing out but action2 has not yet operated on the database. When it attempts to then operate on the database it tries to use the same transaction as action1 but action1 closed that transaction after it hit the timeout error.

**Comment by RahulBirCodes** [comment]
@raymondKelly I was having the same error as you where i had two actions within a Promise.all(). I ended up making the two actions sequential and the problem was fixed.

**Comment by aqrln** [comment]
This is an old comment but I'll respond since the issue surfaced again in my notifications and I don't believe there was an explicit explanation to this specific problem given in this thread before:

> after reading provided solutions and comments, I didn't understand why it works with good old for loop but not with Array.forEach or Array.map and also some people mentioned that using Promise.all solved the issue for their use cases. Can someone clarify the reason and work around for that case? it will be nice for people coming afterwards.

Because `Array.prototype.forEach` accepts a regular function, not an async function. If you provide an async function as an argument, it will execute it as a normal function without awaiting, which will spawn it concurrently.

Code like this in one of the comments above:

```ts
  // DO NOT DO THIS
  deal.activities.forEach(async (activity) => {
    const dealActivity = await tx.crmDealActivity.create()
    // ...
  }
```

is just plainly wrong and doesn't do what you think it does. The transaction callback will finish and commit the transaction before the operations have a chance to execute. You need to await the operations **inside** this transaction callback and not outside of it.

In other words, if you want sequential execution, you should use the `for ... of` loop:

```ts
  for (const activity of deal.activities) {
    const dealActivity = await tx.crmDealActivity.create()
    // ...
  }
```

and if you want concurrent execution, you should use `Promise.all`:

```ts
  await Promise.all(deal.activities.map((activity) => {
    const dealActivity = await tx.crmDealActivity.create()
    // ...
  }))
```

**Don't forget `await` before `Promise.all` or you will ultimately run into the same issue as with `forEach`.**

When querying databases, accessing network or doing any other kind of I/O, **you definitely want concurrent execution whenever you don't have dependencies between the operations, especially in the context of interactive transactions, where sequential execution can lead to timeouts**.

In contrast, **using `forEach` with async functions is almost always a bug in your code, except a specific use case where the following four conditions are true**:
1. You want to spawn concurrent tasks
2. You don't care about their results
3. You don't care when they will finish
4. They can never throw an error (as otherwise you won't be able to catch it and it will crash your application)

It is unfortunately accepted by TypeScript (as functions returning whatever are assignable to the functions returning `void` as long as the parameters match) but you can set up ESLint to reject code like this.

**Comment by patriciorasquetti** [comment]
Alexey answer is correct (Promise.all) , I was implementing it wrong at that time and fixed last year.
Regards.

**Comment by talonx** [comment]
> Ran into this issue today as well – turns out the fix was to convert my `arr.forEach(...)` loop into a good old `for (let i = 0; i < arr.length; i++)` and everything worked like a charm!

Thank you. This saved the day for me.

**Comment by talonx** [comment]
> > Ran into this issue today as well – turns out the fix was to convert my `arr.forEach(...)` loop into a good old `for (let i = 0; i < arr.length; i++)` and everything worked like a charm!
> 
> Thank you. This saved the day for me.

Just to add some more context - I faced this in prod and not in dev, where I was connecting to the prod DB from my local box for a migration, so network latency was definitely higher.

**Comment by dedupely** [comment]
I had the same issue and it was because I was calling queries after the transaction closed. Basically a forgotten await. This probably doesn't help but it could be the case in more complex cases as well, e.g. using set timeouts or anything like that.

**Comment by shadoath** [comment]
When dealing with promises, the behavior of `for`, `map()`, and `forEach()` loops differ significantly due to how they handle asynchronous operations:

### 1. `for` Loop
- **Synchronous Execution**: The `for` loop handles asynchronous code more flexibly because you can use `async/await` directly within the loop.
- **Sequential Execution**: When using `await` inside a `for` loop, each iteration waits for the promise to resolve before moving to the next one, enabling sequential execution of asynchronous tasks.

**Example:**
```javascript
for (const item of items) {
  await doSomethingAsync(item); // Runs one by one
}
```

### 2. `map()`
- **Creates an Array of Promises**: When using `map()` with an asynchronous function, it does not wait for each promise to resolve. Instead, it immediately returns an array of promises.
- **Use with `Promise.all()`**: If you want to run all the promises concurrently and wait for all of them to complete, you typically wrap it in `Promise.all()`.

**Example:**
```javascript
const promises = items.map(item => doSomethingAsync(item));
await Promise.all(promises); // Runs concurrently
```

### 3. `forEach()`
- **No `await` Handling**: The `forEach()` method doesn’t work with `await` directly. If you try to use `await` inside `forEach()`, it won’t pause the loop execution, and all iterations will start their asynchronous operations almost simultaneously.
- **Less Control**: Because it doesn’t handle `await`, it’s generally not suitable for controlling the flow of asynchronous tasks.

**Example:**
```javascript
items.forEach(async (item) => {
  await doSomethingAsync(item); // Won't wait for each iteration
});
```

### Summary
- Use `for` for sequential, controlled execution.
- Use `map()` with `Promise.all()` for concurrent execution.
- Avoid `forEach()` with promises, as it doesn't handle `await` properly.
